src/pudl_archiver/archivers/classes.py

```diff
@@ -392,72 +399,98 @@ async def get_hyperlinks(
         filter_pattern: typing.Pattern | None = None,
         context: str = "text",
-    ) -> list[str]:
+        all_names: bool = False,
+    ) -> dict[str, str | list]:
         """Return all hyperlinks from HTML text.
 
         This is a helper-helper function to perform very basic HTML-parsing functionality.
         It extracts all hyperlinks from an HTML text, and returns those that match
         a specified pattern. This means it can find all hyperlinks that look like
         a download link to a single data resource.
 
+        If a URL appears multiple times with different anchor text, by default we return
+        the anchor text that appears closest to the end of the document. To return all
+        anchor text, run with all_names=True.
+
         Args:
             text: text containing HTML.
             filter_pattern: If present, only return links that contain pattern.
             context: String used in error messages to describe what text was being searched.
+            all_names: default False; if True, return all anchor text found for each link,
+                not just the final one.
         """
         parser = _HyperlinkExtractor()
         parser.feed(text)
 
-        # Filter to those that match filter_pattern
         hyperlinks = parser.hyperlinks
 
+        # Filter to those that match filter_pattern
         if filter_pattern:
-            hyperlinks = {
-                link: name
-                for link, name in hyperlinks.items()
-                if filter_pattern.search(name) or filter_pattern.search(link)
-            }
+            matching_hyperlinks = {}
+            for link, names in hyperlinks.items():
+                if filter_pattern.search(link):
+                    matching_hyperlinks[link] = names
+                    continue
+                # make sure that if we're doing a name based match,
+                # the only names we return are matching names
+                for name in names:
+                    if filter_pattern.search(name):
+                        if link not in matching_hyperlinks:
+                            matching_hyperlinks[link] = []
+                        matching_hyperlinks[link].append(name)
+            hyperlinks = matching_hyperlinks
 
         # Warn if no links are found
         if not hyperlinks:
             self.logger.warning(
-                f"In {context}: the archiver couldn't find any hyperlinks {('that match: ' + filter_pattern.pattern) if filter_pattern else ''}."
+                f"In {context}: the archiver couldn't find any hyperlinks{(' that match: ' + filter_pattern.pattern) if filter_pattern else ''}. "
                 f"Make sure your filter_pattern is correct, and check if the structure of the page is not what you expect it to be."
             )
 
-        return hyperlinks
+        return (
+            hyperlinks
+            if all_names
+            else {link: names.pop() for link, names in hyperlinks.items()}
```

Comment:
(`.pop()` returns the last item in the list, which is the one that appeared closest to the end of the file)